crates/telio-starcast/src/starcast_peer.rs

```diff
@@ -0,0 +1,312 @@
+use boringtun::noise::{Tunn, TunnResult};
+use ipnetwork::IpNetwork;
+use std::{net::SocketAddr, time::Duration};
+use telio_crypto::{PublicKey, SecretKey};
+use telio_proto::DataMsg;
+use telio_task::{
+    io::{wait_for_tx, Chan},
+    task_exec, Runtime, RuntimeExt, Task, WaitResponse,
+};
+use telio_utils::interval;
+use telio_utils::{telio_log_error, telio_log_warn};
+use telio_wg::uapi::Peer;
+use tokio::{net::UdpSocket, sync::mpsc::error::SendTimeoutError, time::Interval};
+use x25519_dalek::{PublicKey as PublicKeyDalek, StaticSecret};
+
+const MAX_PACKET: usize = 2048;
+
+#[derive(Debug, thiserror::Error)]
+/// Custom `UdpProxy` error
+pub enum Error {
+    /// `StarcastPeer` Not Configured Error
+    #[error("Starcast peer is not configured")]
+    NotConfigured,
+    /// `StarcastPeer` `IoError`
+    #[error("Starcast peer IoError")]
+    IoError(#[from] std::io::Error),
+    /// Task execution failed
+    #[error(transparent)]
+    Task(#[from] telio_task::ExecError),
+    /// IP network failed
+    #[error("Failed to parse IP network")]
+    IpNetworkError(#[from] ipnetwork::IpNetworkError),
+    /// Send failed
+    #[error(transparent)]
+    Send(#[from] SendTimeoutError<(PublicKey, DataMsg)>),
+    /// Tunnel creation failed
+    #[error("Tunnel creation failed {0}")]
+    TunnNewFail(&'static str),
+}
+
+// TODO this should probably be moved to a higher level file
+
+/// The actual starcast peer containing the task which runs the wait loop for this whole component.
+pub struct StarcastPeer {
+    task: Task<State>,
+}
+
+/// Starcast peer configuration.
+pub struct Config {
+    /// Public key of the starcast virtual peer.
+    public_key: PublicKey,
+    /// Port number through which the virtual starcast peer shall communicate with WireGuar.
+    wg_port: u16,
+}
+
+/// State for the starcast peer runtime where all the "working" members are stored and operated on.
+struct State {
+    /// Allowed IPs of the starcast peer.
+    allowed_ips: Vec<IpNetwork>,
+    /// Interval at which the tunnel timers will be updated.
+    wg_timer_tick_interval: Interval,
+    /// Channel for sending decapsulated packets to the multicaster and receiving packets for encapsulation from it.
+    channel: Chan<Vec<u8>>,
+    /// Local socket for sending and receiving packets from WireGuard.
+    socket: Option<UdpSocket>,
+    /// WireGuard socket address.
+    wg_sock_addr: Option<SocketAddr>,
+    /// WireGuard encryption tunnel for encapsulation/decapsulation of packets.
+    tunnel: Option<Tunn>,
+    /// Flag indicating that there are packets in the encryption tunnel internal buffer that need to be processed
+    /// before new packets can be received from the socket and decapsulated.
+    packets_present_in_tunnel: bool,
+    // TODO need to recheck if there are no issues with these buffers:
+    /// Buffer for outbound packets.
+    sending_buffer: Vec<u8>,
+    /// Buffer for inbound packets.
+    receiving_buffer: Vec<u8>,
```

Comment:
I might be incorrect as you have multiple arms, but maybe it's posible to use a singular buffer

Reply:
Handling inbound packets needs two buffers. One for reading from socket and another one to decapsulate into.